pre_commit_hooks/ros_include_guard.py

```diff
@@ -0,0 +1,154 @@
+      self.expected = ''.join(self.tokens)
+    self.nolint = allow_nolint and ('NOLINT' in self.tokens)
+
+
+class ClosingDirective(Directive):
+
+  def prepare(self, macro_name, allow_nolint):
+    self.tokens = re.split(R'\b', self.text)
+    self.expected = '#endif  // {}'.format(macro_name)
+    self.nolint = allow_nolint and ('NOLINT' in self.tokens)
+
+
+class IncludeGuard:
+
+  def __init__(self):
+    self.ifndef = OpeningDirective()
+    self.define = OpeningDirective()
+    self.endif  = ClosingDirective()
+    self.pragma = False
+
+  def items(self):
+    yield self.ifndef
+    yield self.define
+    yield self.endif
+
+  def is_none(self):
+    return any(item.is_none() for item in self.items())
+
+  def mismatch(self):
+    return any(item.mismatch() for item in self.items())
+
+  def overwrite(self, lines):
+    return any(item.overwrite(lines) for item in self.items())
+
+  def prepare(self, macro_name, allow_nolint):
+    for item in self.items(): item.prepare(macro_name, allow_nolint)
+
+
+def get_include_guard_info(lines):
+
+  guard = IncludeGuard()
+  for line, text in enumerate(lines):
+    if text.startswith('#pragma once'):
+      guard.pragma = True
+    if text.startswith('#ifndef') and guard.ifndef.is_none():
+      guard.ifndef.update(line, text)
+    if text.startswith('#define') and guard.define.is_none():
+      guard.define.update(line, text)
+    if text.startswith('#endif'):
+      guard.endif.update(line, text)
+  return guard
+
+
+def get_parts_after(parts: list, targets: set):
+  result = []
+  for part in reversed(parts):
+    if part in targets:
+      break
+    result.append(part)
+  return reversed(result)
+
+
+def get_include_guard_macro_name(filepath: pathlib.Path):
+  targets = {'include', 'src', 'test'}
+  for path in filepath.parents:
+      if path.joinpath('package.xml').exists():
+          parts = get_parts_after(filepath.relative_to(path).parts, targets)
+          return '__'.join(parts).replace('.', '_').upper() + '_'
+  return None
+
+
+def main(argv=None):
+
+  parser = argparse.ArgumentParser()
+  parser.add_argument('filenames', nargs='*', help='Filenames to fix')
+  parser.add_argument('--allow-nolint', action='store_true', help='Allow nolint comment')
+  args = parser.parse_args(argv)
+
+  return_code = 0
+  for filename in args.filenames:
+    filepath = pathlib.Path(filename)
+    macro_name = get_include_guard_macro_name(filepath)
+    if macro_name:
+      lines = filepath.read_text().split('\n')
+      guard = get_include_guard_info(lines)
+      if guard.pragma:
+        continue
+      if guard.is_none():
+        return_code = 1
+        print('No include guard in {}'.format(filepath))
+        continue
+      guard.prepare(macro_name, args.allow_nolint)
+      if guard.mismatch():
+        return_code = 1
+        print('Fix include guard in {}'.format(filepath))
+        guard.overwrite(lines)
+        filepath.write_text('\n'.join(lines))
+        continue
```

Comment:
It's great if you could write some supplemental comments to briefly explain the steps so that we can imagine the flow chart in our head.

Reply:
Commented. https://github.com/tier4/pre-commit-hooks-ros/pull/32/commits/f1f2a94e7df7f37a63660bd4d7753954b7b19497